include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
+             *                       Calling synchronization within the functor is allowed.
              */
-            template<typename T_Species, typename T_Functor, typename T_AreaMapperFactory>
-            HINLINE void forEach(T_Species&& species, T_Functor functor, T_AreaMapperFactory const& areaMapperFactory)
+            template<typename T_Acc, typename T_Functor>
+            DINLINE void operator()(T_Acc const& acc, T_Functor&& unaryFunctor) const
+            {
+                if constexpr(numWorkers > frameSize)
+                {
+                    PMACC_CASSERT_MSG(
+                        __number_of_workers_must_be_a_multiple_of_framesize,
+                        (numWorkers % frameSize) == 0);
+                }
+
+                auto const& superCell = m_particlesBox.getSuperCell(m_superCellIdx);
+                uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+
+                // Information used to know if a particle within a frame is active without checking the multiMask.
+                auto frameConditionData = T_Order::createCtx(forEachParticleInFrame);
+
+                // end kernel if we have no particles
+                if(numPartcilesInSupercell == 0)
+                    return;
+
+                // get starting frame
+                auto framePtrCtx = forEachParticleInFrame(
+                    [&](lockstep::Idx const idx)
+                    {
+                        auto frame = T_Order::getFirst(m_particlesBox, m_superCellIdx);
+
+                        if constexpr(std::is_same_v<Reverse, T_Order>)
+                        {
+                            if(frame.isValid())
+                                frameConditionData[idx]
+                                    = m_particlesBox.getSuperCell(m_superCellIdx).getSizeLastFrame();
+                        }
+
+                        /* Same as numFramesToSkip, variable is required to workaround an nvcc compiler bug.
+                         * Without redefinition within the lambda the variable will be undefined.
+                         */
+                        constexpr uint32_t skipNFrames = (numWorkers + frameSize - 1u) / frameSize;
+                        if constexpr(skipNFrames > 1)
+                        {
+                            uint32_t const frameIdx = idx / frameSize;
+                            /* select N-th (N=frameIdx) frame from the list */
+                            for(uint32_t i = 1; i <= frameIdx && frame.isValid(); ++i)
+                            {
+                                frame = T_Order::next(m_particlesBox, frame);
+                            }
+                            if constexpr(std::is_same_v<Reverse, T_Order>)
+                                frameConditionData[idx] = frameIdx != 0 ? frameSize : frameConditionData[idx];
+                        }
+                        return frame;
+                    });
+
+                bool isOneFrameValid = true;
```

Comment:
Why `one`? To me it seems more like a `isCurrentFrameValid`

Reply:
A worker can have multiple frames depending on the number of virtual threads the worker is representing.
We need to know if one of these frames a worker is holding is valid.

Reply:
offline discussed @sbastrakov suggested: `hasUnprocessedFrames` or `hasMoreWork`